compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETransferStridedAccessPattern.cpp

```diff
@@ -0,0 +1,377 @@
+    SmallVector<OpFoldResult> &dmaSizes, SmallVector<OpFoldResult> &dmaStrides,
+    SmallVector<OpFoldResult> &circularDmaOffsets,
+    SmallVector<OpFoldResult> &circularDmaSizes,
+    SmallVector<OpFoldResult> &circularDmaStrides) {
+  IRRewriter rewriter(ctx);
+
+  // Make copies of L3 original sizes and strides which will be needed later
+  // when creating new L2 addressing.
+  SmallVector<OpFoldResult> l3OrigSizes = dmaSizes;
+  SmallVector<OpFoldResult> l3OrigStrides = dmaStrides;
+
+  FailureOr<size_t> isCombinable =
+      isL3AddressingCombinable(dmaOffsets, dmaSizes, dmaStrides);
+  if (failed(isCombinable)) {
+    return emitError(rewriter.getUnknownLoc())
+           << "failed to get dim position for combination";
+  }
+  size_t dimForCombine = isCombinable.value();
+
+  // Generate L3 side new source offsets/sizes/strides.
+  // Example: [[0, 0, 0] [2, 32, 32] [32, 128, 1]] will become
+  // [[0, 0] [32, 64] [128, 1]] after the first and the innermost dims are
+  // combined.
+  DenseSet<size_t> excludeDims = {dimForCombine};
+  dmaOffsets = copyExcludeDims(dmaOffsets, excludeDims);
+  dmaStrides = copyExcludeDims(dmaStrides, excludeDims);
+
+  std::optional<SmallVector<int64_t>> maybeSizes =
+      getConstantIntValues(l3OrigSizes);
+  std::optional<SmallVector<int64_t>> maybeStrides =
+      getConstantIntValues(l3OrigStrides);
+  if (!maybeSizes.has_value() || !maybeSizes.has_value()) {
+    return emitError(rewriter.getUnknownLoc())
+           << "failed to get original source sizes / strides.";
+  }
+  SmallVector<int64_t> sizeVals = maybeSizes.value();
+  SmallVector<int64_t> strideVals = maybeStrides.value();
+
+  int64_t innerDimTotal = strideVals.back() * sizeVals.back();
+  int64_t newInnerSize = sizeVals[dimForCombine] * innerDimTotal;
+
+  size_t lastIndex = l3OrigSizes.size() - 1;
+  excludeDims.insert(lastIndex);
+  dmaSizes = copyExcludeDims(dmaSizes, excludeDims);
+  dmaSizes.push_back(getAsIndexOpFoldResult(ctx, newInnerSize));
+
+  // Generate L2 side new target offsets/sizes/strides.
+  SmallVector<OpFoldResult> newCircularOffsets(l3OrigSizes.size(),
+                                               rewriter.getIndexAttr(0));
+  circularDmaOffsets = newCircularOffsets;
+
+  circularDmaSizes = copyExcludeDims(l3OrigSizes, excludeDims);
+  circularDmaSizes.push_back(
+      getAsIndexOpFoldResult(ctx, sizeVals[dimForCombine]));
+  circularDmaSizes.push_back(getAsIndexOpFoldResult(ctx, innerDimTotal));
+
+  // Function to create new strides for NpuCircularDmaCpyNdOp.
+  auto getNewL2Strides = [&](SmallVector<int64_t> values) {
+    SmallVector<OpFoldResult> res = {getAsIndexOpFoldResult(ctx, 1)};
+    int64_t initial = values.back();
+    // Leave out one dimension for insertion afterwards
+    for (size_t i = values.size() - 2; i > 0; i--) {
+      initial *= values[i];
+      res.push_back(getAsIndexOpFoldResult(ctx, initial));
+    }
+    return llvm::to_vector(llvm::reverse(res));
+  };
```

Comment:
Take the following dma ops for example:
```
%45 = amdaie.npu.circular_dma_cpy_nd %8([0] [2048] [1], [] [] [])
%46 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[0, 0, 0, %41] [4, 2, 32, 32] [4096, 32, 128, 1])

```
The logic is to create L2 side strides from the innermost dimension, and then reverse the vector to have the final order. The new L2 side strides always start with [1], and should have the same number of dimensions as the original L3 side source addressing. The next dimensions are calculated by the logic `initial *= l3OrigSizes[i]`.

The `initial` means the innermost continuous elements which is `l3OrigSizes[-1]* l3OrigStrides.back[-1]` (the implementation omit l3OrigStrides[-1] because l3OrigStrides[-1] == 1). The combined elements are now continuous on L3 side, but should have a strided addressing on L2 side, the stride should be `initial * l3OrigSizes[-2]`. So after this iteration, the strides are [1, 32 * 32].

Same logic for the next iteration, the strides become [1, 32 * 32, 32 * 32 * 2] = [1, 1024, 2048]. After reversion, it's [2048, 1024, 1]. At last insert the stride for the position of the combined dimension, which is l3OrigStride[dimForCombine], and get final strides [2048, 32, 1024, 1].

Let me know if this is the correct logic to get the L2 side strides, or if there's a better way to calculate this.

Reply:
> The combined elements are now continuous on L3 side, but should have a strided addressing on L2 side, the stride should be initial * l3OrigSizes[-2].

Yeah, I think the core idea is good: i.e. that the strides should be created as if the original L3 was contiguous and then rearranged based on which dimension(s) are combined with the innermost one.

However, I do think this needs extensive tests to ensure correctness as this will otherwise lead to hard-debug numerical errors in the future. So, it would be good to create a standalone utility function that takes in a set of static offsets/sizes/strides and produces the new static L3 and L2 offsets/sizes/strides, so that it can be tested standalone (ctest, not lit) on a lot of different cases, see for example: https://github.com/nod-ai/iree-amd-aie/blob/main/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp